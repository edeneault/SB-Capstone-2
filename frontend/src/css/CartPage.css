.card-cart {
  min-height: 10rem;
}
