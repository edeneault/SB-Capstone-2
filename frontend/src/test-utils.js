import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";

// create a customRender that wraps the UI in a memory Router

const MemoryRouterWithInitialRoutes = ({ children, initialRoutes }) => {
  return <MemoryRouter initialEntries={initialRoutes}>{children}</MemoryRouter>;
};

const customRender = (ui, options) => {
  const initialRoutes =
    options && options.initialRoutes ? options.initialRoutes : ["/"];

  return render(ui, {
    wrapper: (args) =>
      MemoryRouterWithInitialRoutes({
        ...args,
        initialRoutes,
      }),
    ...options,
  });
};

// re-export everything
export * from "@testing-library/react";

// override render method
export { customRender as render };
