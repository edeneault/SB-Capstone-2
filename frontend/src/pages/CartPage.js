import React, { useEffect } from "react";
import { Link } from "react-router-dom";
import { useDispatch, useSelector } from "react-redux";
import {
  Container,
  Row,
  Col,
  ListGroup,
  Image,
  Form,
  Button,
  Card,
} from "react-bootstrap";
import Message from "../components/Message";
import MetaData from "../components/MetaData";
import { addToCart, removeFromCart } from "../actions/cartActions";

import "../css/CartPage.css";

const CartPage = ({ match, location, history }) => {
  const productId = match.params.id;

  // uses the query portion of the url, in this case "?qty=1" split on "=" will returns array with index 0 and 1.
  // The quantity is equal to the value at index 1. Or the default quantity of 1.
  const qty = location.search ? Number(location.search.split("=")[1]) : 1;

  const dispatch = useDispatch();

  const cart = useSelector((state) => state.cart);
  const { cartItems } = cart;

  useEffect(() => {
    if (productId) {
      dispatch(addToCart(productId, qty));
    }
  }, [dispatch, productId, qty]);

  const removeFromCartHandler = (id) => {
    dispatch(removeFromCart(id));
  };

  const checkoutHandler = () => {
    history.push("/login?redirect=shipping");
  };

  return (
    <>
      <MetaData title='cart page' />
      <Container fluid className='px-5 pt-3'>
        <Row>
          <Col md={6}>
            <i className='fas fa-shopping-cart fa-2x px-1 py-3 text-black'></i>
            <h1 className='shopping-cart-title d-inline py-2'>
              {" "}
              Shopping Cart
            </h1>
            {cartItems.length === 0 ? (
              <Message>
                Your Cart is empty <Link to='/'>Go Back</Link>
              </Message>
            ) : (
              <ListGroup variant='flush'>
                {cartItems.map((item) => (
                  <ListGroup.Item key={item.product}>
                    <Row>
                      <Col md={2}>
                        <Image src={item.image} alt={item.name} fluid rounded />
                      </Col>
                      <Col md={3}>
                        <Link to={`/products/${item.product}`}>
                          {item.name}
                        </Link>
                      </Col>
                      <Col md={2}>${item.price}</Col>
                      <Col md={2}>
                        <Form>
                          <Form.Control
                            as='select'
                            value={item.qty}
                            onChange={(e) =>
                              dispatch(
                                addToCart(item.product, Number(e.target.value)),
                              )
                            }
                          >
                            {[...Array(item.countInStock).keys()].map((x) => (
                              <option key={x + 1} value={x + 1}>
                                {x + 1}
                              </option>
                            ))}
                          </Form.Control>
                        </Form>
                      </Col>
                      <Col md={2}>
                        <Button
                          type='button'
                          variant='light'
                          onClick={() => removeFromCartHandler(item.product)}
                        >
                          <i className='fas fa-trash'></i>
                        </Button>
                      </Col>
                    </Row>
                  </ListGroup.Item>
                ))}
              </ListGroup>
            )}
          </Col>
        </Row>
        <Row>
          <Col md={12} className='py-3'>
            <Card className='card-cart'>
              <ListGroup variant='flush'>
                <ListGroup.Item>
                  <h2>
                    Subtotal (
                    {cartItems.reduce((acc, item) => acc + item.qty, 0)}) items
                  </h2>
                  $
                  {cartItems
                    .reduce((acc, item) => acc + item.qty * item.price, 0)
                    .toFixed(2)}
                </ListGroup.Item>

                <ListGroup.Item>
                  <Col className='d-grid'>
                    <Button
                      type='button'
                      className='btn-block'
                      disabled={cartItems.length === 0}
                      onClick={checkoutHandler}
                    >
                      Proceed to Checkout
                    </Button>
                  </Col>
                </ListGroup.Item>
              </ListGroup>
            </Card>
          </Col>
        </Row>
      </Container>
    </>
  );
};

export default CartPage;
