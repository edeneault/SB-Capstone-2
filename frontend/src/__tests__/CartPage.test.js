import React from "react";
import { Provider } from "react-redux";
import configureMockStore from "redux-mock-store";
import thunk from "redux-thunk";
import { render } from "../test-utils";
import { initialState } from "../store-mock";
import CartPage from "../pages/CartPage.js";
import { createMemoryHistory, createLocation } from "history";

const mockStore = configureMockStore([thunk]);

describe("test", () => {
  const history = createMemoryHistory();
  const path = `/cart/:id`;

  const match = {
    isExact: false,
    path,
    url: path.replace(":id", "6125d04db949259bdec4de3d"),
    params: { id: "6125d04db949259bdec4de3d" },
  };
  const location = createLocation(match.url);

  test("renders CartPage without crashing", () => {
    const store = mockStore(initialState);

    const { getByText } = render(
      <Provider store={store}>
        <CartPage match={match} location={location} history={history} />
      </Provider>,
      { initialRoutes: [`/cart`] },
    );

    expect(getByText(/Shopping Cart/i)).toBeInTheDocument();
  });

  test("matches snapshot", function () {
    const store = mockStore(initialState);

    const { asFragment } = render(
      <Provider store={store}>
        <CartPage match={match} location={location} history={history} />
      </Provider>,
      { initialRoutes: [`/cart`] },
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
