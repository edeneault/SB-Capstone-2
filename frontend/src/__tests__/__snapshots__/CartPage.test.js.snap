// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="px-5 pt-3 fade-in container-fluid"
  >
    <div
      class="row"
    >
      <div
        class="col-md-6"
      >
        <i
          class="fas fa-shopping-cart fa-2x px-1 py-3 text-black"
        />
        <h1
          class="shopping-cart-title d-inline py-2"
        >
           Shopping Cart
        </h1>
        <div
          class="list-group list-group-flush"
        >
          <div
            class="list-group-item"
          >
            <div
              class="row"
            >
              <div
                class="col-md-2"
              >
                <img
                  alt="Lavazza - Espresso - Medium Roast"
                  class="img-fluid rounded"
                  src="https://res.cloudinary.com/edeneault-cloudinary/image/upload/v1630783263/eShop/sqiwu5ummm4y1to3cp5o.jpg"
                />
              </div>
              <div
                class="col-md-3"
              >
                <a
                  href="/products/6125d04db949259bdec4de3c"
                >
                  Lavazza - Espresso - Medium Roast
                </a>
              </div>
              <div
                class="col-md-2"
              >
                $9.99
              </div>
              <div
                class="col-md-2"
              >
                <form
                  class=""
                >
                  <select
                    class="form-control"
                  >
                    <option
                      value="1"
                    >
                      1
                    </option>
                    <option
                      value="2"
                    >
                      2
                    </option>
                    <option
                      value="3"
                    >
                      3
                    </option>
                    <option
                      value="4"
                    >
                      4
                    </option>
                  </select>
                </form>
              </div>
              <div
                class="col-md-2"
              >
                <button
                  class="btn btn-light"
                  type="button"
                >
                  <i
                    class="fas fa-trash"
                  />
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="row"
    >
      <div
        class="py-3 col-md-12"
      >
        <div
          class="card-cart card"
        >
          <div
            class="list-group list-group-flush"
          >
            <div
              class="list-group-item"
            >
              <h2>
                Subtotal (1) items
              </h2>
              $9.99
            </div>
            <div
              class="list-group-item"
            >
              <div
                class="d-grid col"
              >
                <button
                  class="btn-block btn btn-primary"
                  type="button"
                >
                  Proceed to Checkout
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
